### DISCUSSION TITLE
Implementing Louvain Community Detection

### DISCUSSION THREAD

**Discussion by z3y50n** [networkx/networkx#4669] [Algorithms]
So I was looking at existing PRs that tried to implement Louvain inside the networkx libary. Namely #1090 is pretty much a copy of [python-louvain](https://github.com/taynaud/python-louvain) library with a wrapper for the output. #1092 seems to contain more unique code (although it follows the same reference as the `python-louvain` library).

One part of the discussion in #951 that will decide a lot in my opinion regarding implementing Louvain from scratch or just adding support for `python-louvain` library is the response object of the algorithm. Pretty much everyone agreed on using a `list` of `sets` as it gives more freedom for other uses while `python-louvain`\`s method returns a dictionary where keys are the nodes and values are the community they belong to.

I am starting this discussion since I would like to finally implement Louvain inside networkx in the coming GSoC (either from scratch or from existing tools) and what I mentioned above is a small recap I managed to do regarding the work that has already been made but is on halt right now. I would love to hear your opinions on which approach should be followed, since I am starting to make a plan on how the workflow will go. 

Right now it seems to me that the best approach is to try and integrate the existing library inside networkx and to use some helper functions to modify the results according to our needs, but maybe there are some limitations on `python-louvain` that I am not aware of and someone else here is and therefore I should try to implement my own version of Louvain.

As a final note I have found this [paper](https://arxiv.org/pdf/1410.1237.pdf) that implements a parallel version of Louvain and it is worth to take a look in order to optimize the serial implementation of Louvain (when that happens).

**Comment by dschult** [comment]
This looks like a good synopsis of the current state of those tickets and the references.
I guess the next item of interest is the current API choices of the rest of the community detection routines in NetworkX.
For example, do we return a dict of nodes to their community name? Or do we return a list of community sets? Or are there some functions that return one and some return the other? We certainly have code to convert from one form to the other, but I'm not sure how visible that conversion code is.  To fold the new algorithm into our current interface we need to understand the current interface. I hope it is fairly uniform, but given the long piecemeal history it might not be. :}

**Reply by z3y50n** [reply]
Thanks for your feedback. You can see an initial draft of my [application](https://drive.google.com/file/d/1efZ5bIbJc0bmWjucwKxTz6U_e1xc6N2H/view?usp=sharing).
I have concluded that a list/iterable of sets is the best option since most community algorithms implemented inside NetworkX already use this structure. In my proposal I state that we could create a class to handle the output of communities according to the user needs.

**Reply by z3y50n** [reply]
Also please let me know if I should post my application on the networkx google group as well.

**Reply by MridulS** [reply]
No need to post the application on the google group.
The draft looks good!

Also don't forget to sign up on https://summerofcode.withgoogle.com and officially submit the application there :)
